src/manifolds/FixedRankMatrices.jl

```diff
@@ -287,16 +284,18 @@ singular values and $U$ and $V$ are shortened accordingly.
 retract(::FixedRankMatrices, ::Any, ::Any, ::PolarRetraction)
 
 function retract!(
-    ::FixedRankMatrices{M,N,k},
+    ::FixedRankMatrices{m, n, k},
     q::SVDMPoint,
     p::SVDMPoint,
     X::UMVTVector,
     ::PolarRetraction,
-) where {M,N,k}
-    s = svd(p.U * Diagonal(p.S) * p.Vt + (p.U * X.M * p.Vt + X.U * p.Vt + X.U * X.Vt))
-    q.U .= s.U[:, 1:k]
-    q.S .= s.S[1:k]
-    q.Vt .= s.Vt[1:k, :]
+) where {m, n, k}
+    QU, RU = qr([p.U X.U])
+    QV, RV = qr([p.Vt' X.Vt'])
+    T = svd(RU * [diagm(p.S) + X.M I; I zeros(k, k)] * RV')
```

Comment:
The expression within the `svd(...)` allocates 6 matrices. It would be preferable to perform the multiplication blockwise to avoid allocating so many intermediates. e.g. something like
```suggestion
    @views begin
        RU11 = RU[:, 1:k]
        RU12 = RU[:, (k + 1):(2 * k)]
        RV11 = RV[:, 1:k]
        RV12 = RV[:, (k + 1):(2 * k)]
    end
    tmp = RU11 .* p.S' .+ RU12
    mul!(tmp, RU11, X.M, true, true)
    tmp2 = tmp * RV11'
    mul!(tmp2, RU11, RV12', true, true)
    T = svd(tmp2)
```
would allocate just two temporaries. (perhaps you can think of more descriptive variable names)

Reply:
(definitely check my math on this one)